### PR TITLE
fix(workflow): preserve reviewer loop retry history

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

- Add a script-owned `reviewer_loop_history.v1` JSON payload to the existing automated reviewer-loop summary comment.
- Preserve retry entries across summary updates and expose unavailable states instead of treating unreadable history as zero retries.
- Update retrospective protocol and runner guidance to prefer durable reviewer-loop history for exact retry metrics.

Closes #1243

## Links

- Spec: `docs/specs/developments/20260716104000_1243-preserve-reviewer-loop-retry-history/1_1243-preserve-reviewer-loop-retry-history_specs.md`
- Plan: `docs/specs/developments/20260716104000_1243-preserve-reviewer-loop-retry-history/2_1243-preserve-reviewer-loop-retry-history_implementation-plan.md`
- Smoke test runbook: `docs/testing/workflow/1243-preserve-reviewer-loop-retry-history.smoke-test.md`

## Document Quality Gate

- Scope checked against approved spec and implementation plan for #1243.
- Existing readiness, CI, tracker, label, and merge-authority semantics left unchanged.
- Durable history is appended to the existing single script-owned summary comment rather than creating duplicate timeline comments.

## Parser-Risk Checklist

| Edge | Coverage |
| --- | --- |
| Missing history block | `history_first_entry_*` tests create iteration 1 and zero retries |
| Malformed JSON | `history_malformed_*` tests return unavailable with `malformed_history` |
| Wrong schema | `history_wrong_schema_reason` returns unavailable with `unknown_schema` |
| Empty entries | `history_empty_entries_append_count` appends the first entry |
| Multiple blocks | `history_latest_block_preserved` chooses the latest block |
| Fence boundary | `history_fence_boundary_append_count` keeps adjacent Markdown out of JSON |
| Prior unavailable | `history_prior_unavailable_preserved` preserves the prior reason |
| Needs rerun | `history_needs_rerun_result` records `needs_rerun` |
| Duplicate same SHA | `history_same_sha_duplicate_appends` appends without overwrite |
| Comment API read failure | `history_read_failure_*` preserves `comment_read_failed` |
| Duplicate summary comments | Existing Area 13 summary update-in-place tests remain green |

## Validation

- `bash -n scripts/development-workflow/pr-review-loop.sh scripts/development-workflow/tests/test-pr-review-loop.sh`
- `bash scripts/development-workflow/tests/test-pr-review-loop.sh` - 306 passed, 0 failed
- `shellcheck --severity=warning -x scripts/development-workflow/pr-review-loop.sh scripts/development-workflow/tests/test-pr-review-loop.sh`
- `python3 scripts/lint/workflow-shell-guard-lint.py --base-ref origin/develop`
- `npx markdownlint-cli2 "docs/workflow/development-workflow/protocols/06-retrospective-protocol.md" ".claude/agents/retrospective.md" ".cursor/agents/retrospective.md" ".codex/skills/workflow-retrospective/SKILL.md" ".agents/skills/retrospective/SKILL.md" "docs/testing/workflow/1243-preserve-reviewer-loop-retry-history.smoke-test.md" "CHANGELOG.md"`
- `find docs/specs/developments docs/testing/workflow -name "*.md" -print0 | xargs -0 python3 scripts/lint/markdown-heuristic-lint.py CHANGELOG.md`
- `bash scripts/lint/check-changelog-duplicate-headers.sh CHANGELOG.md`
- `git diff --check`

## Smoke Test

Deterministic shell harness coverage was used for this tooling change instead of creating a disposable live reviewer PR. The runbook scenarios are covered by mocked PR comment fixtures in `scripts/development-workflow/tests/test-pr-review-loop.sh`, including multi-iteration append, single clean run, malformed/unavailable history, and duplicate-summary update-in-place behavior.

---

> [!NOTE]
> **Medium Risk**
> Changes core `pr-review-loop.sh` summary comment behavior and retrospective metrics sourcing; risk is mitigated by extensive harness tests and fail-closed unavailable states when history cannot be read.
> 
> **Overview**
> Adds durable **`reviewer_loop_history.v1`** JSON to the existing script-owned **Automated Reviewer Loop Summary** comment so each loop run appends an iteration record (result, blockers, head SHA, timestamps) instead of losing retry context on in-place comment updates.
> 
> **`pr-review-loop.sh`** gains history extract/append/render helpers, merges prior history when PATCHing the single summary comment (including release-guard skips), surfaces **`history_status: unavailable`** when JSON is missing or corrupt, and posts a summary on **`needs_rerun`** so fix-pushed cycles count toward retrospective metrics. **`06-retrospective-protocol.md`** and retrospective agents/skills now treat that payload as authoritative for automated-reviewer retry loops (`max(entries.length - 1, 0)`), with legacy timestamp heuristics only as fallback.
> 
> **`test-pr-review-loop.sh`** adds Area 10b coverage for schema, append, malformed history, comment-read failures, and summary selector behavior; **CHANGELOG** documents the change (#1243).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ed3dd2b892ce011765f4ad4efb1951f08c757e3c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## why_safe_to_merge

```json
{
  "why_safe_to_merge": {
    "scope": "Single workflow-tooling PR for reviewer-loop summary history and retrospective metric guidance; no product runtime, database, auth, or deployment behavior changed.",
    "tests": "Reviewer-loop harness passed with 306 tests, including comment-read failure recovery, available-history preservation, parser edge cases, duplicate summary update-in-place, and Bugbot review-thread regression coverage.",
    "reviewer_outcome": "Step 7a internal review approved after fixing hidden comments API pipeline failure; PR-Agent clean; Cursor Bugbot clean on commit ed3dd2b with zero unresolved review threads and no late-arriving threads.",
    "ci_outcome": "pr-ci-loop.sh reported green for PR #1261: 8 checks, 0 failing, 0 pending, including Markdown lint, policy, PR-Agent review, ShellCheck static analysis, workflow test harnesses, and Cursor Bugbot.",
    "rollback_or_cleanup_risk": "Rollback is limited to reverting the workflow script/test/doc changes; no persistent data migration or external configuration change. Post-merge cleanup only deletes the implementation branch/worktree."
  }
}
```

### CODE REVIEW

**Comment by lhpaul** [comment]
### Step 7a Internal Review Gate Summary

**PR type**: Implementation (two-pass)
**Effective reviewer set**: codex
**Skipped reviewers**: none

**Pass 1 (Spec Compliance)**

- codex: APPROVED after 1 fix cycle. Found and fixed a comment-fetch failure handling gap so unreadable existing summary comments now emit `comment_read_failed` instead of starting false available history.

**Pass 2 (Code Quality)**

- codex: APPROVED after 1 fix cycle. Replaced hidden pipeline failure handling with `set -o pipefail` around the comments API read, added explicit optional metadata failure handling, and extended harness coverage.

**Verdict**: APPROVED
All passes approved at commit `038adee`.

**Comment by lhpaul** [comment]
bugbot run

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), bugbot (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** bugbot

*Posted automatically by `pr-review-loop.sh`.*

File: scripts/development-workflow/pr-review-loop.sh [hunk 1]
```diff
@@ -5850,12 +6132,25 @@ EOF
...
-            | .id // empty
+            | {id: (.id // ""), body: (.body // "")}
           '
     )" \
-      || { echo "WARN: failed to fetch existing summary comments for PR ${pr_number}; will create a new comment" >&2; _existing_comment_id=""; }
+      || {
+        echo "WARN: failed to fetch existing summary comments for PR ${pr_number}; will create a new comment with unavailable history" >&2
+        _existing_comment_record=""
+        _existing_comment_body="$(reviewer_loop_history_unavailable_stub_body comment_read_failed)"
+      }
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/pr-review-loop.sh:6149]: Fixed in commit `ed3dd2b`: comment selection now patches the newest script-owned summary while preserving the newest available history body after transient read-failure stubs, with harness coverage for the orphaned-history case.

**Comment by lhpaul** [comment]
bugbot run

**Comment by lhpaul** [comment]
## /run-epic PR Disposition

- Scope source: explicit /run-items 1184,1185,1187,1242,1243
- Item: #1243 - tooling: preserve reviewer-loop retry history for retrospectives
- PR: #1261
- Reviewed head SHA: `ed3dd2b`
- Reviewer result: clean
- Blocking findings: 0
- Advisory findings: 0
- Risk: medium
- Merge authority: User confirmed mayMerge=true and maxRisk=medium for explicit /run-items scope.
- Final decision: merge_approved

### Risk Reasons

- workflow script or harness change: scripts/development-workflow/pr-review-loop.sh
- workflow guidance change: docs/workflow/development-workflow/protocols/06-retrospective-protocol.md

### Invocation Policy

- Original command: `$run-items 1184, 1185, 1187, 1242, 1243 until merged; confirm mayMerge=true maxRisk=medium`
- Copy-paste equivalent: `$run-items 1184,1185,1187,1242,1243 --delegate-review true --may-merge true --max-risk medium --base develop`
- Confirmation: accepted by user in current session

| Field | Recommended | Selected | Effective |
| --- | --- | --- | --- |
| mayStartBacklog | true | true | true |
| delegateReview | true | true | true |
| mayMerge | true | true | true |
| maxRisk | medium | medium | medium |
| base | develop | develop | develop |

### Checkpoint Policy

- Field source: none
- Pending applicable checkpoints: 0

| Item | Stage | Domain | State | Reason | Required action |
| --- | --- | --- | --- | --- | --- |

### Advisory Decisions

None.

### Verification Evidence

- Labels: ready-for-human-review, ready-for-regression
- CI result: green
- Reviewer summary present: true
- Unresolved thread count: 0
- PR merge state: CLEAN
- Issue state: open
- Project status: Development in Review

### Protocol Deviations

| Action | Impact | Mitigation |
| --- | --- | --- |
| non-force branch history preservation | created a normal merge commit instead of force-pushing amended PR history | remote branch history preserved; final branch contains pushed commit and follow-up fixes |
